pynocular/backends/memory.py

```diff
@@ -0,0 +1,280 @@
+
+    This backend stores records in memory. It translates SQLAlchemy expressions into
+    Python operations. It should only be used in tests.
+    """
+
+    def __init__(self, records: Optional[Dict[str, List[Dict[str, Any]]]] = None):
+        """Initialize a SQLDatabaseModelBackend
+
+        Args:
+            records: Optional map of table name to list of records to bootstrap the
+                in-memory database
+
+        """
+        super().__init__()
+        self.records = records or defaultdict(list)
+        # Serial primary key generator
+        self._pk_generator = itertools.count(start=1)
+
+    def _set_primary_key_values(
+        self,
+        config: DatabaseModelConfig,
+        record: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        """Set default values on a record for the primary keys
+
+        Args:
+            config: DatabaseModelConfig instance that contains references to a table and
+                columns that can be used to build queries suited to the backend.
+            record: The record to update
+
+        Returns:
+            updated record
+
+        """
+        for primary_key in config.primary_keys:
+            value = (
+                next(self._pk_generator)
+                if isinstance(primary_key.type, Integer)
+                else str(uuid4())
+            )
+            record.setdefault(primary_key.name, value)
+
+        return record
+
+    @staticmethod
+    def _update_db_managed_fields(
+        config: DatabaseModelConfig,
+        record: Dict[str, Any],
+        fetch_on_create: bool = False,
+        fetch_on_update: bool = False,
+    ) -> Dict[str, Any]:
+        """Update record values for db managed fields
+
+        Args:
+            config: DatabaseModelConfig instance that contains references to a table and
+                columns that can be used to build queries suited to the backend.
+            record: The record to update
+            fetch_on_create: Flag that controls whether the db managed field will be
+                updated if it has the option `fetch_on_create=True`. Defaults to False.
+            fetch_on_update: Flag that controls whether the db managed field will be
+                updated if it has the option `fetch_on_update=True`. Defaults to False.
+
+        Raises:
+            NotImplementedError: if a field sets fetch_on_create or fetch_on_update to
+                true but its type is not supported
+
+        Returns:
+            updated record
+
+        """
+        for name in config.db_managed_fields:
+            field = config.fields[name]
+            if (fetch_on_create and field.field_info.extra.get("fetch_on_create")) or (
+                fetch_on_update and field.field_info.extra.get("fetch_on_update")
+            ):
+                if field.type_ == datetime:
+                    record[name] = datetime.utcnow()
+                else:
+                    raise NotImplementedError(field.type_)
```

Comment:
There is a bug here since primary IDs also have a `fetch_on_create` on them so they fall into this else statement